### PR TITLE
Optimize API Calls by Enabling Conditional Request Based on User Setting

### PR DESCRIPTION
**Pull request by SubodhDahal**
## **User description**
In the current implementation,  we are always asking for improved code. But we may not be interested in the actual code suggestion rather just the text suggestion. So I added a new option to disable getting improved code. 

___

## **Type**
Enhancement

___

## **Description**
- This PR optimizes the API calls by enabling conditional requests based on user settings.
- The 'existing code' and 'improved code' are now included in the PR code suggestions and YAML keys only if the user has enabled this feature.
- This change reduces unnecessary API calls, improves performance, and reduces token usage.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Conditional Inclusion of 'existing code' and 'improved code' in YAML 

 Keys</code></dd></summary>
<hr>
      
pr_agent/algo/utils.py

- Modified the `keys` list in `try_fix_yaml` function to conditionally <br>  include 'existing code:' and 'improved code:' based on the user <br>  settings.

    </details>
  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/621/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_code_suggestions_prompts.toml</strong><dd><code>Conditional Inclusion of Improved Code in PR Code Suggestions Prompts</code></dd></summary>
<hr>
      
pr_agent/settings/pr_code_suggestions_prompts.toml

- Updated the CodeSuggestion class to conditionally include <br>  'existing_code' and 'improved_code' based on the user settings.

    </details>
  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/621/files#diff-81979f64eddc8f93a5c6ea83ff1e7ba92e551acaecd310c606255161e914c2ea">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pr_code_suggestions.py</strong><dd><code>Conditional Inclusion of Improved Code in PR Code Suggestions</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/tools/pr_code_suggestions.py

- Added a new key 'include_improved_code' to the settings dictionary.
- Modified `_prepare_pr_code_suggestions` and <br>  `push_inline_code_suggestions` functions to conditionally include <br>  'existing_code' and 'improved_code' based on the user settings.

    </details>
  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/621/files#diff-b57ba775e741d6f80bc4f8154b71330c011dae0ac43f3d0197e785b3e6b7117b">+11/-7</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

**Comment by mrT23** [comment]
I removed this option, which existed, yesterday.

I don't see the use case.
You don't want to see the suggestion, don't open the collapsible.

There are also mistakes and inconsistencies in the implementation

File: pr_agent/settings/pr_code_suggestions_prompts.toml [hunk 1]
```diff
@@ -79,17 +81,21 @@ code_suggestions:
...
   relevant_lines_end: 12
 {%- else %}
   existing_code: |-
     def func1():
+{%- endif %}
   relevant_lines_start: 12
   relevant_lines_end: 12
+{%- if include_improved_code %}
   improved_code: |-
     ...
```
Comment (mrT23) [inline pr_agent/settings/pr_code_suggestions_prompts.toml:100]: this section is implemented wrongly

**Comment by SubodhDahal** [comment]
Through extensive use of GPT-4 for my codebase, I have discovered that the suggested improvements are incorrect far more frequently than one might expect, even when the textual suggestions seem valid. Given this reality, it made little sense to waste time and resources reviewing incorrect code changes. To streamline the process and minimize unnecessary costs, I extended the existing `include_improved_code` option to exclude the request for improved code altogether.

File: pr_agent/settings/pr_code_suggestions_prompts.toml (hunk 1 above)
Comment (SubodhDahal) [reply pr_agent/settings/pr_code_suggestions_prompts.toml:100]: Could you elaborate, please? It looks correct to me as I am including the pattern for "improved_code" only when the option is set.

**Comment by meicale** [comment]
@CodiumAI-Agent /review

**Comment by mrT23** [comment]
@SubodhDahal 
- the 'review' feature already provides suggestions without "code". you can use it
- the implementation you provided is wrong. The order of the 'ifs' is incorrect and inconsistent. print the actual prompt that is generated and you will see.
- I don't like to add flags if I don't think it will be applicable to many people, and not specific ones. especially in a crucial place like the prompt.


I thank for for the contribution, but I am not going to merge it. If you are kin about this specific implementation, you can use it in your local hosting

thank you

**Comment by SubodhDahal** [comment]
I understand @mrT23. I am going to close the MR.

**Pull request closed without merging**